File: abc/049/049c.java
import java.util.*;

public class Main {
	public static void main(String[] args){
    Scanner sc = new Scanner(System.in);
    String s = sc.next();
    String dream = "dream";
    String dreamer = "dreamer";
    String erase = "erase";
    String eraser = "eraser";

    String answer = "YES";
    int i = s.length() - 1;
    while (i >= 0) {
      if (s.charAt(i) == 'm') {
        if (i - 4 >= 0 && Objects.equals(s.substring(i - 4, i + 1), dream)) {
          i -= 5;
          continue;
        } else {
          answer = "NO";
          break;
        }
      } else if (s.charAt(i) == 'e') {
        if (i - 4 >= 0 && Objects.equals(s.substring(i - 4, i + 1), erase)) {
          i -= 5;
          continue;
        } else {
          answer = "NO";
          break;
        }
      } else if (s.charAt(i) == 'r') {
        if (i - 6 >= 0 && Objects.equals(s.substring(i - 6, i + 1), dreamer)) {
          i -= 7;
          continue;
        } else if (i - 5 >= 0 && Objects.equals(s.substring(i - 5, i + 1), eraser)) {
          i -= 6;
          continue;
        } else {
          answer = "NO";
          break;
        }
      } else {
        answer = "NO";
        break;
      }
    }

    System.out.println(answer);
  }
}

File: abc/049/049c.kt
fun main(args: Array<String>) {
  var s = readLine()!!

  var i = s.length

  while (i >= 0) {
    // erase or dream
    if (i >= 5) {
      val sub5 = s.substring(i - 5, i)
      if (sub5 == "erase" || sub5 == "dream") {
        s = s.substring(0, i - 5)
        i -= 5
        continue
      }
    }
    // eraser
    if (i >= 6) {
      val sub6 = s.substring(i - 6, i)
      if (sub6 == "eraser") {
        s = s.substring(0, i - 6)
        i -= 6
        continue
      }
    }
    // dreamer
    if (i >= 7) {
      val sub7 = s.substring(i - 7, i)
      if (sub7 == "dreamer") {
        s = s.substring(0, i - 7)
        i -= 7
        continue
      }
    }

    if (s == "") {
      println("YES")
      return
    } else {
      println("NO")
      return
    }
  }
}
